File: 04-stack/00020-Valid_Parentheses/00020-valid_parentheses.py
# 20: Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/


class Solution:
    # SOLUTION
    def isValid (self, s: str) -> bool:
        parenthesis = []

        for c in s:
            if c=='{':
                parenthesis.append('}')
            elif c=='[':
                parenthesis.append(']')
            elif c=='(':
                parenthesis.append(')')
            else:
                if len(parenthesis)==0 or c!=parenthesis.pop():
                    return False
        
        return len(parenthesis)==0


if __name__ == "__main__":
    o = Solution()

    # INPUT
    s = "()[]{}"

    # OUTPUT
    result = o.isValid(s)
    print(result)

File: 04-stack/00020-Valid_Parentheses/00020-valid_parentheses.java
// 20: Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/


import java.util.Stack;

class Solution {
    // SOLUTION
    public boolean isValid (String s) {
        Stack<Character> parentheses = new Stack<>();

        for (var c : s.toCharArray()) {
            switch(c) {
                case '{': parentheses.push('}'); break;
                case '[': parentheses.push(']'); break;
                case '(': parentheses.push(')'); break;
                default:
                    if (parentheses.isEmpty() || c!=parentheses.pop())
                        return false;
            }
        }

        return parentheses.isEmpty();
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String s = "()[]{}";

        // OUTPUT
        var result = o.isValid(s);
        System.out.println(result);
    }
}